### ISSUE TITLE
schedule

### ISSUE THREAD

**Issue by calin-darie** [calin-darie/gentle-glow-onyx-boox#6]
- [x] demand - react with a ❤ to this comment to let me know I should work on this feature. Maybe also donate and add a 🍹reaction to the ❤
- [x] spec ready for first development
- [x] technical tasks defined
- [x] first app store release
- [ ] tested and approved in app store

The kindle Oasis might be a source of inspiration: choose a profile to automatically default to at night. 

Maybe gradually shift between the day and the night profile around sunset? 

Automatically figure out when the sunset given the device's time zone?

Extend this to multiple automatic profile switches throughtout the day?

https://www.youtube.com/watch?v=pM2bLJx7cAg

**Comment by arooni** [comment]
just created a duplicate issue but a schedule would be amazing !!!!

**Comment by fraslin** [comment]
I love the idea of a schedule as switch back and forth every day (and perfect app for the Boox!)

**Comment by blchinezu** [comment]
This would be awesome! It's the only thing I'm missing from my old Kobo Aura One. 

The way it works on Kobo is to define the hour at which you go to sleep. It then gradually increases the warm light starting an hour (or so) earlier than defined. What I like about this is that you get the wanted warmth and don't even notice the change while reading.

I'm not sure when it starts dropping the warm light though. I'm assuming 8h after the defined hour? Don't know.

It does have a light sensor though so managing the screen brightness is an easy task for Kobo. For Boox we'll probably need to manage brightness through schedules too.

I'm not saying this is the way to implement it but I thought the info can't hurt.

From my point of view, having 2 schedules (for night & morning) and gradual brightness/warmth change would be all I can ask for.

**Comment by calin-darie** [comment]
Spec draft:
You will get the possibility of setting a time of day when the current profile is auto-selected. A time picker and a will be available to the right of the name edit box.
The scheduled transitions will be smoothed automatically, with a transition speed depending on two factors:
1. difference between brightness and warmth of the profiles you're transitioning between e.g. going from (30/100 brightness, 50/100 warmth) to (15/100 brightness, 100/100 warmth) will be slower than going from (30/100 brightness, 50/100 warmth) to (20/100 brightness, 75/100 warmth)
2. time allowed for the transition. e.g. if you configure tranition to the Sunset profile at 18:00 then to the Night profile at 18:05, the transition might happen more quickly than if you configure transition to Sunset at 18:00 but transition to Night at 18:30.

**Comment by calin-darie** [comment]
Changing the current light configurationwill affect all scheduled transitions to this configuration.
There will be an option, though, to just adjust the light without affecting any configuration. For this purpose, plus resetting to the last known onyx slider configuration , a dropdown menu will be added at the right of the profile radio buttons.

**Comment by calin-darie** [comment]
A new screen will be added, to edit the schedule. The interface will be similar to the alarm clock list view
```
7:00 (On) Dawn ✏❌
9:00 (Off) ✏❌
18:00 (On) Sunset ✏❌
21:00 (On) Night ✏❌
➕
```
Off/On switch lets you choose a profile
```
✏ takes you to the profile
❌ deletes the time
➕ adds a new time
```

**Comment by calin-darie** [comment]
Sorry for the delay on this, I'm two months behind schedule :( => earliest possible release date is 1st of May

**Comment by thehilsnz** [comment]
Can't wait for this feature! It's the only thing I could ask for - I always switch between two presets, one for day and one for night. Thanks for your work on this.

**Comment by ahanuban** [comment]
Very much looking forward to this. Thank you!

**Comment by donxepe** [comment]
Someone give this dev a beer! 🍺

**Comment by cuong21951** [comment]
> Sorry for the delay on this, I'm two months behind schedule :( => earliest possible release date is 1st of May

Did this finish ?

**Comment by cuong21951** [comment]
> Sorry for the delay on this, I'm two months behind schedule :( => earliest possible release date is 1st of May

I added when the device booted I load all the shedule and find the closet shedule and manually invoke it. I also change to RTC_WAKEUP. Do you want me to create a PR ?

**Comment by calin-darie** [comment]
> I added when the device booted I load all the shedule and find the closet shedule and manually invoke it. I also change to RTC_WAKEUP. Do you want me to create a PR ?

Hello, sure, I finally have some time again for Gentle Glow so please open a PR

**Comment by calin-darie** [comment]
Actually @cuong21951 turns out Ii had a small unpushed commit to restore alarms on boot, see https://github.com/calin-darie/gentle-glow-onyx-boox/compare/restore-alarms?expand=1

**Comment by sergmetelin** [comment]
Hi @calin-darie, did you manage to make it work? I think automatic scheduling is one of the big features missing in Boox Go Color 7, for example. Palma has auto-brightness/warmth, but Go is missing it :(

**Comment by jakemauer** [comment]
Just chiming in to say that timers would be lovely to have. I'm using a Boox 6 Go as a smart home dashboard with Home Assistant and would love the backlight to be on at night and then go off late in the evening. The real ideal would be some sort of programmatic control that I could control via Home Assistant, but in lieu of that, timers would be awesome.

**Comment by thehilsnz** [comment]
Super excited to see the Schedule feature in the latest update. Unfortunately it crashes every time I try and set it on my Go Colour 7. I’ve tried reinstalling, restarting etc. I can open the schedule UI, it crashes if I try to enable it but then when I reopen it’s saved the enable which is great. But if I try and add a new schedule it lets me enter a time but then crashes and hasn’t saved anything. I’m hoping the schedule will at least work with the default settings for now. Thanks again for working on this!

**Comment by calin-darie** [comment]
@thehilsnz sorry to hear that

Is it only the schedule that makes it crash? Can you use the brightness & warmth controls normally?

Do you still have your old devices? How does this new version perform on them?

**Comment by thehilsnz** [comment]
Yep, just the schedule that crashes, the core lighting features work fine. Also, the schedule didn’t trigger last night. 

In a stroke of bad timing I sold off my other 3 Boox devices last week, so I can’t test on them!

**Comment by delorfin** [comment]
Can confirm crashes on Go Color 7. I can't add a time; the app crashes after selecting a time in the system date picker.

**Comment by delorfin** [comment]
For the schedule, it’d be great to be able to set sunset and sunrise times depending on the location, so the times would be automatically updated throughout the year.

If requesting the location permission is undesirable, the location could be specified with lat/long coordinates. There are free APIs that provide sunset and sunrise (and also dusk and dawn) times: https://sunrise-sunset.org/api, https://sunrisesunset.io/api/. To get the coordinates, people could go to https://www.latlong.net/ and enter their city.

**Comment by jakemauer** [comment]
Reporting in that my first test of the scheduler on my Boox 6 Go works. I'm going to let it run tonight while I have Home Assistant open and report back.

**Comment by calin-darie** [comment]
@thehilsnz , @delorfin  your reports seem to hint at trouble with time math and setting the alarm, so probably Android-related, not Boox-related

I'll test on my Android 14 phone with a dumbed down app and hopefully I can reproduce the bug. Otherwise, I'll be stuck and I'll need a hand to get at least a stack trace.

**Comment by thehilsnz** [comment]
Your latest update works! No more crashes on the Go Color 7 and the scheduling works perfectly. Thank you!

**Comment by thehilsnz** [comment]
One question - is it hard/possible to include the option to set the front light off on schedule as well as the presets? The lowest I can save a preset is to a brightness of 1 so that doesn’t quite work for what I’m after.

**Comment by calin-darie** [comment]
Yes, @thehilsnz . In the scheduler UI, the switch next to the time is for the light on / off state, not to activate the schedule entry. So when you add a new schedule entry the default is lights off at that time. If you want to skip an entry, delete it.

**Comment by thehilsnz** [comment]
Oh yes, so it does, thank you!
